swift/extractor/main.cpp

```diff
@@ -3,15 +3,58 @@
 #include <stdlib.h>
 
 #include <swift/Basic/LLVMInitialize.h>
+#include <swift/FrontendTool/FrontendTool.h>
+
+#include "Extractor.h"
+
+using namespace std::string_literals;
+
+class Observer : public swift::FrontendObserver {
+ public:
+  explicit Observer(const codeql::Configuration& config) : config{config} {}
+
+  void performedSemanticAnalysis(swift::CompilerInstance& instance) override {
+    codeql::Extractor extractor{config, instance};
+    extractor.extract();
+  }
+
+ private:
+  const codeql::Configuration& config;
+};
+
+static std::string getenv_default(const char* envvar, const std::string& def) {
+  if (const char* var = getenv(envvar)) {
+    return var;
+  }
+  return def;
+}
 
 int main(int argc, char** argv) {
+  if (argc == 1) {
+    /// TODO: print usage
+    return 1;
+  }
+  /// The frontend can be called in different modes, we are only interested
+  /// in -frontend mode
+  /// TODO: filter out at the tracer level
+  if ("-frontend"s != argv[1]) {
+    return 0;
+  }
   PROGRAM_START(argc, argv);
-  if (auto trapDir = getenv("CODEQL_EXTRACTOR_SWIFT_TRAP_DIR")) {
-    std::string file = trapDir;
-    file += "/my_first.trap";
-    if (std::ofstream out{file}) {
-      out << "answer_to_life_the_universe_and_everything(42)\n";
+
+  codeql::Configuration configuration{};
+  configuration.trapDir = getenv_default("CODEQL_EXTRACTOR_SWIFT_TRAP_DIR", ".");
+  configuration.sourceArchiveDir = getenv_default("CODEQL_EXTRACTOR_SWIFT_SOURCE_ARCHIVE_DIR", ".");
+  std::vector<const char*> args;
+  for (int i = 1; i < argc; i++) {
+    if ("-frontend"s == argv[i]) {
+      continue;
     }
+    args.push_back(argv[i]);
   }
-  return 0;
+  configuration.frontendOptions = args;
+  Observer observer(configuration);
+  int frontend_rc = swift::performFrontend(args, "swift-extractor", (void*)main, &observer);
+  llvm::llvm_shutdown();
```

Comment:
Optional: Might be nice to wrap this in [`wil::scope_exit`](https://github.com/microsoft/wil/wiki/RAII-resource-wrappers#wilscope_exit).

Reply:
FWIW, LLVM has [llvm_shutdown_obj](https://github.com/llvm/llvm-project/blob/09c2b7c35af8c4bad39f03e9f60df8bd07323028/llvm/include/llvm/Support/ManagedStatic.h#L118), though we don't need it as it's shutdown by the `PROGRAM_START`.

Reply:
does that mean we don't need this line then?

btw, I would rather use [gsl::finally](https://github.com/microsoft/GSL/blob/main/include/gsl/util#L89-L96) than wil::scope_exit (as I'd rather pull in the `gsl` library then `wil`).